services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java

```diff
@@ -228,4 +224,19 @@ protected boolean isTemporarilyDisabledByBruteForce(AuthenticationFlowContext co
         }
         return false;
     }
+
+    private void setInvalidUser(AuthenticationFlowContext context) {
+        context.getEvent().error(Errors.USER_NOT_FOUND);
+        Response challengeResponse;
+        if (getDefaultChallengeMessage().equals(Messages.INVALID_USERNAME) && context.getRealm().isLoginWithEmailAllowed()) {
```

Comment:
How about if signature of "getDefaultChallengeMessage()" is changed to "getDefaultChallengeMessage(AuthenticationFlowContext context)" ? Then in UsernameForm, you will be able to retrieve the realm and return either INVALID_USERNAME or INVALID_USERNAME_OR_EMAIL . This means that this "if/else" in invalidUser won't be needed as it will be entirely the responsibility of authenticator to return the error. WDYT?

Reply:
Yes, I've considered this possibility, but the `getDefaultChallengeMessage` with parameter `context` would be used only with `UsernameForm`. So I thought, this way would be cleaner. 
But, if you think, it'll be better with parameter, I'll remake it :)